### PR TITLE
[BUU] Add Turbo for dynamic page loading

### PR DESCRIPTION
**Pull request by dacook**
### What? Why?
After removing StimulusReflex, the page doesn't feel as responsive. Can we use the Turbo framework to make it snappy again??

  * [x] problem: filters are appearing twice when there are form errors??! ⚠️. Maybe due to https://github.com/hotwired/turbo/issues/132, which is meant to be fixed, so I'm probably missing something. In anycase I don't understand what conditions would cause it...
  * [x] problem: body classes get cleared sometimes. Dunno why, but I've removed dependency on them.
  * [x] "Clear search" button. Dunno why it doesn't use Turbo
  * [ ] Implement Streams format responses, so we can save loading time. (This is the equivalent to StimulusReflex, and is the ideal solution for https://github.com/openfoodfoundation/openfoodnetwork/issues/11987)
  * [ ] Lazy-load the products-content frame separately on initial load.
  * [ ] problem: The loading spinner disappears too early, before the page has updated.

In summary, yes, although there's some fine-tuning to do!

### What should we test?
All the things!! 🫠 (almost)
Please report on the general look and feel too, any opinions are appreciated.

* Page load: shows products as expected (still no loading indicator yet sorry)
* Filtering: by search term, products, category
* Pagination: 
   * choose number per page
   * browse pages
* Submit changes (with a filter and a page selected)
   * Products
   * Variants
   * Form errors (eg blank product name)
* New variant
   * Form errors (eg blank unit)
* Product actions work as before
   * Edit image 
   * Edit record (warn about unsaved changes)
      * Save edits
   * Delete
   * Clone (warn about unsaved changes)
* Variant actions: these work the same as product, so no need to test again.
* New product (warn about unsaved changes)

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [x] Feature toggled

The title of the pull request will be included in the release notes.

#### Documentation updates
We gotta update the wiki

### CODE REVIEW

File: spec/system/admin/products_v3/products_spec.rb
Comment (rioug) [inline spec/system/admin/products_v3/products_spec.rb:156]: Is this supposed to be commented out ?

File: app/views/admin/products_v3/_content.html.haml
Comment (rioug) [inline app/views/admin/products_v3/_content.html.haml:1]: I think there are some library to debug morphing which should be able to tell you if it's working. See https://github.com/leastbad/radiolabel It's for `CableReady` though so I don't know if that's going to work with turbo

**Review by rioug — APPROVED** [review]
Nice job :+1:  exciting to see Turbo in action ! 

I believe Turbo stream uses websocket under the hood, so depending on what you are using it for we might run into the same issue we had with Stimulus Reflex.

File: spec/system/admin/products_v3/products_spec.rb
Comment (dacook) [reply spec/system/admin/products_v3/products_spec.rb:156]: Sort of. Currently it says "1 product***s*** found for your search criteria". So technically this should be `pending`. 
It must be time to go ahead and fix that..

File: app/views/admin/products_v3/_content.html.haml
Comment (dacook) [reply app/views/admin/products_v3/_content.html.haml:1]: Thanks, I did end up using the code from [hotwire-debug](https://github.com/beflagrant/hotwire-debug/tree/main), and might be installing that soon. This logs turbo events and could show the morph event I think.

**Comment by dacook** [comment]
> I believe Turbo stream uses websocket under the hood, so depending on what you are using it for we might run into the same issue we had with Stimulus Reflex.

You're right, it depends how we use it:

https://turbo.hotwired.dev/handbook/streams
> These elements can be delivered to the browser synchronously as a classic HTTP response, or asynchronously over transports such as webSockets, SSE, etc

We should be able to replace Reflex actions with good old Controller actions, by adding the `turbo_stream` format, something like this: (caveat, I haven't got it working smoothly yet)

```ruby
      respond_to do |format|
        format.html do
          render "index", locals: { producers:, categories:, products: @products, flash: }
        end
        format.turbo_stream do |format|
          # I think we can DRY this up with a clever pattern too.
          render turbo_stream: turbo_stream.replace("products-content", partial: "content",
            locals: { products: @products, flash: })
        end
      end
```

**Comment by dacook** [comment]
Sorry for the mass-review requests, but I thought it worth having everybody aware of this one.

File: config/routes/spree.rb
Comment (mkllnk) [inline config/routes/spree.rb:62]: Clone wars! What a funny side effect. Are we sure we want to keep the pre-loading? Would that apply to pagination as well? The faster user experience comes at the cost of more server load and unnecessary requests.

**Review by mkllnk — APPROVED** [review]
Very cool that you got it working now!

I wonder what that learning curve means to contributors and myself. I always need to look things up in the documentation and this didn't seem easy. It seems to require some special knowledge to know what magic is happening. It's not self-explanatory.

File: config/routes/spree.rb
Comment (dacook) [reply config/routes/spree.rb:62]: In this case, the pagination got implemented with a form submit, not links, so it doesn't apply there. But it does apply to the "discard changes" link, which reloads the product list.
We really don't want it firing off lots of expensive requests without good cause! 

In other cases, it's quite nice, eg when you click on New or Edit product, it can take you there instantly.

The handbook discusses some ways you can tweak this behaviour (https://turbo.hotwired.dev/handbook/drive#prefetching-links-on-hover), but probably the simplest and safest is to disable pre-fetch in the admin interface entirely for now. I will do that on the next revision.

Comment (dacook) [reply config/routes/spree.rb:62]: Really, I think pre-fetch should only be used on cacheable pages (and hopefully Turbo is smart enough not to re-fetch them). So it would be most useful on the frontend.

**Comment by dacook** [comment]
> I wonder what that learning curve means to contributors and myself. I always need to look things up in the documentation and this didn't seem easy. It seems to require some special knowledge to know what magic is happening. It's not self-explanatory.

That is a very important question. I think this debug gem would help a lot: https://github.com/beflagrant/hotwire-debug/tree/main . It shows you what frames are being updated, and logs all the turbo events so you can see what is or isn't happening.

**Comment by dacook** [comment]
I found a bug:
* Product actions work as before
   * Edit record (warn about unsaved changes)
      * ❌ Save edits: the "Changes that you made might not be saved" warning from the previous page appears. 😬 

So I had to revert a couple of things and downgrade some links to basic links. I think we can solve this better if Turbo would implement something like a `before-frame-visit` event. (will look into this....)

**Comment by dacook** [comment]
As we have two approvals, with a minor fix added, I'm going to move to Test Ready.

**Review by chahmedejaz — APPROVED** [review]
Great work 👍🏻
It was a great opportunity to learn something now. Excited to use it as well😁

**Review by anansilva — APPROVED** [review]

File: config/routes/spree.rb
Comment (mkllnk) [reply config/routes/spree.rb]: > pre-fetch should only be used on cacheable pages

Totally agree. We should choose an opt-in approach.

**Review by mkllnk — APPROVED** [review]

**Comment by RachL** [comment]
Hi @dacook I did a first run which seem fine. I will merge this to unblock work and we can still fix if there are stuff that I have missed.
FYI I didn't test

> New product (warn about unsaved changes)
As the new product page does not warn about unsaved change?

Also:

> choose number per page

This was a bit slow: first to load the 100 products, then to change the "Showing 1 to 100".  But I think that's what you meant by:

> problem: The loading spinner disappears too early, before the page has updated.

Let me know if you need any help on triaging the fine tuning left.

**Comment by dacook** [comment]
Thanks Rachel, 

> New product (warn about unsaved changes)

Sorry that wasn't at all clear, my bad. I meant if you had unsaved changes, then clicked "New product", it should warn you. 
But never mind, I just checked and it worked :D

> choose number per page
> This was a bit slow: first to load the 100 products, then to change the "Showing 1 to 100".

Interesting, I just tried on au_staging and was surprised how quick it loaded the first page of 100 (about 6 seconds), particularly compared to the old products screen (about 14 seconds, half of which is browser-side rendering). 
It's not a perfect comparison because they were different products (I'm not sure why but they are sorted differently), but still  I think it shows how the new architecture can perform a lot better!!

> Let me know if you need any help on triaging the fine tuning left.

Thanks, I was going to go through my notes and raise/update issues, but ran out of steam today! I'll raise in Slack.

**Pull request merged**
